src/pinyin.cpp

```diff
@@ -0,0 +1,513 @@
+#include "pinyin.h"
+
+#include <algorithm>
+#include <map>
+#include <string>
+#include <cuchar>
+#include <unordered_map>
+#include <vector>
+
+namespace pinyin
+{
+
+/* data is referenced from https://github.com/rime/rime-terra-pinyin (which is an IME setup for chinese input) */
```

Comment:
Probably need to figure out if there is legal concern or consequence from directly embedding this LGPLv3 licensed data into the source code of this CC BY-SA 3.0 licensed project without a library boundary.

Reply:
Not sure why they chose LGPL for two yaml files which are not library source files. Perhaps we could turn these files into a library under the LPGL license and link them here but that sounds kinda ridiculous.

Also it seems that repo does not include some pronunciations of simplified characters (https://github.com/rime/rime-terra-pinyin/pull/49#issuecomment-1418266742) so we probably should use a different source anyway.

Perhaps we could use Unihan_Readings.txt from https://www.unicode.org/Public/UCD/latest/ucd/Unihan.zip which seems to have a compatible license (https://www.unicode.org/license.txt) but this files seems to have inconsistent format and needs normalization before we can use it.

Reply:
To be honest I have no idea how I could make this work. Even if I generate data through other means, or by reading a dictionary and manually typing in the data, the result would mostly be the same, because the pronunciation of a character does not change no matter how the work is done...

I'll see if I can come up with alternative methods for this

Reply:
> Perhaps we could use the kHanyuPinyin field from Unihan_Readings.txt in https://www.unicode.org/Public/UCD/latest/ucd/Unihan.zip which seems to have a compatible license (https://www.unicode.org/license.txt) but this files seems to have inconsistent format and needs normalization before we can use it.

Yeah, I found that as well, but I might just format it like before (pinyin:characters map), I don't know if I should put that into a separate file in the third_party folder and attach the unicode's license to it

Reply:
> To be honest I have no idea how I could make this work. Even if I generate data through other means, or by reading a dictionary and manually typing in the data, the result would mostly be the same, because the pronunciation of a character does not change no matter how the work is done...

Please disregard that, I think if we attach the license for Unihan or Unicode it should be fine, it seems like most similar work are derived from the Unihan database that you just mentioned